extensions/tags/src/Query/TagFilterGambit.php

```diff
@@ -55,23 +55,23 @@ protected function constrain(Builder $query, $rawSlugs, $negate)
     {
         $slugs = explode(',', trim($rawSlugs, '"'));
 
-        $query->where(function (Builder $query) use ($slugs, $negate) {
-            foreach ($slugs as $slug) {
-                if ($slug === 'untagged') {
-                    $query->whereIn('discussions.id', function (Builder $query) {
-                        $query->select('discussion_id')
-                            ->from('discussion_tag');
-                    }, 'or', ! $negate);
-                } else {
-                    $id = $this->tags->getIdForSlug($slug);
-
-                    $query->whereIn('discussions.id', function (Builder $query) use ($id) {
-                        $query->select('discussion_id')
-                            ->from('discussion_tag')
-                            ->where('tag_id', $id);
-                    }, 'or', $negate);
+        $query
+            ->distinct()
+            ->leftJoin('discussion_tag', 'discussions.id', '=', 'discussion_tag.discussion_id')
+            ->where(function (Builder $query) use ($slugs, $negate) {
+                foreach ($slugs as $slug) {
+                    if ($slug === 'untagged' && ! $negate) {
+                        $query->orWhereNull('discussion_tag.tag_id');
+                    } elseif ($slug === 'untagged' && $negate) {
+                        $query->orWhereNotNull('discussion_tag.tag_id');
+                    } elseif ($id = $this->tags->getIdForSlug($slug)) {
+                        $query->orWhere(
+                            'discussion_tag.tag_id',
+                            $negate ? '!=' : '=',
+                            $id
+                        );
+                    }
```

Comment:
Let's try this instead:
```suggestion
        $query
            ->distinct()
            ->join('discussion_tag', function (JoinClause $join) use ($slugs, $negate) {
                $join
                    ->on('discussions.id', '=', 'discussion_tag.discussion_id')
                    ->where(function (\Illuminate\Database\Query\JoinClause $join) use($slugs, $negate) {
                        foreach ($slugs as $slug) {
                            if ($slug === 'untagged' && ! $negate) {
                                $join->orWhereNull('discussion_tag.tag_id');
                            } elseif ($slug === 'untagged' && $negate) {
                                $join->orWhereNotNull('discussion_tag.tag_id');
                            } elseif ($id = $this->tags->getIdForSlug($slug)) {
                                $join->orWhere(
                                    'discussion_tag.tag_id',
                                    $negate ? '!=' : '=',
                                    $id
                                );
                            }
                        }
                    });
            });
```

Reply:
Alas that doesn't result in a significant speedup. But if I change `->join()` into `->leftJoin()` then the request times drop dramatically. Seems to be doing what we want!

As said; I'm not exactly proficient with Eloquent but can't the `->where()` be replaced with `->on()`? That way `\Illuminate\Database\Query\JoinClause` isn't needed. That would be a bit more readable/cleaner because in the end it's not really a `WHERE` but a part of the `ON` part of that `LEFT JOIN`.

Great work!